java/arcs/core/allocator/Arc.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.allocator
+
+import arcs.core.common.ArcId
+import arcs.core.host.ArcState
+import arcs.core.host.ArcState.Deleted
+import arcs.core.host.ArcState.Error
+import arcs.core.host.ArcState.Indeterminate
+import arcs.core.host.ArcState.NeverStarted
+import arcs.core.host.ArcState.Running
+import arcs.core.host.ArcState.Stopped
+import kotlinx.coroutines.CompletableDeferred
+
+/**
+ * Represents an instantiated Arc running on one or more [ArcHost]s. An [Arc] can be stopped
+ * via [Arc.stop], and best-effort state changes can be listened for via [onArcStateChange].
+ *
+ * NOTE: that currently [onArcStateChange] listeners are not persistent across service restarts,
+ * therefore if an [ArcHost] is running in a [Service], or on another device, and they restart,
+ * you will not receive updates from those hosts.
+ *
+ * TODO: add some mechanism to detect host crashes and re-register state change listeners.
+ */
+data class Arc internal constructor(
+    val arcId: ArcId,
+    private val allocator: Allocator,
+    private var arcStateInternal: ArcState = NeverStarted
+) {
+    private val arcStateChangeHandlers = mutableListOf<(ArcState) -> Unit>()
+    private val arcStatesByHostId = mutableMapOf<String, ArcState>()
+
+    /**
+     *  The current running state of an Arc. This is computed by computing the dominant state
+     *  across all [ArcHost]s in an [Arc], where dominant ordering is defined by the ordering in
+     *  [ArcState], for example, [Error] is greater than [Running], so if one [ArcHost] is in
+     *  state [Error], and another is in state [Running], the overall state of the [Arc]
+     *  is considered to be [Error].
+     **/
+    var arcState: ArcState
+        get() = arcStateInternal
+        private set(state) {
+            if (arcStateInternal != state) {
+                arcStateInternal = state
+                fireArcStateChange()
+            }
+        }
+
+    private suspend fun onArcStateChange(handler: (ArcState) -> Unit) {
+        if (arcStateChangeHandlers.isEmpty()) {
+            registerChangeHandlerWithArcHosts()
+        }
+        arcStateChangeHandlers += handler
+        handler(arcState)
+    }
+
+    private suspend fun onArcStateChangeFiltered(stateToFilter: ArcState, handler: () -> Unit) {
+        if (arcState == stateToFilter) {
+            handler()
+        } else {
+            onArcStateChange {
```

Comment:
Is it okay that this listener will get called every time the state changes (even after the `stateToFilter` is reached), but line 68 will only result in the handler being called once?